### DISCUSSION TITLE
Need help with AOT and Trimming, CompiledBindings

### DISCUSSION THREAD

**Discussion by dbriard** [AvaloniaUI/Avalonia#11115] [Q&A] [answered]
Hi all, 
I'm new to these topics and would appreciate guidance on the following points:

I've managed to compile a simple sample based on the UNO tutorial on WASM Fixed SIMD (https://platform.uno/blog/safari-16-4-support-for-webassembly-fixed-width-simd-how-to-use-it-with-c/)
The sample consists of one button with two benchmarks: one with SIMD and one without.

The tutorial highlights that Fixed SIMD can only function with AOT compilation, which means that if I intend to use Fixed SIMD, I must compile with AOT. Additionally, the benchmark shows that C# code appears to perform significantly faster with AOT compilation, even in the absence of SIMD.

1) To perform AOT compilation for a WASM project, it appears that I need to use the command "dotnet publish -c Release" instead of publishing from the UI. It seems that when I publish from the UI, the project is not AOT compiled, as indicated by the benchmark results that show similar performance to the Release version (90ms version 7 ms for non SIMD code, and 300 ms vs 1 ms for SIMD).
![image](https://user-images.githubusercontent.com/3595443/233967154-0680a5f9-3c65-49f5-9921-cb0310dda837.png)

**I am curious why AOT compilation does not occur when I publish from the UI?**

**As my project becomes more complex, would enabling AOT and trimming in the root project (the Avalonia browser project) is enough, or do I need to enable these features for all projects in my solution?**
![image](https://user-images.githubusercontent.com/3595443/233969513-0e058e83-ccc3-4f98-bd66-d505f4956b56.png)

2) In addition to AOT compilation for speed, I would also like to enable Trimming to reduce the size of my project. To test this feature, I created another simple sample consisting of a ViewModel and a view with a combobox that displays a list of fonts. When a user selects a font, the list displays all the icons associated with it. This sample was created by updating an Avalonia Cross platform app from preview4 to preview6 and making a few additions to the view and ViewModel.

...but that do not work as expected... and my ViewModel is almost empty.

For example:
![image](https://user-images.githubusercontent.com/3595443/233968589-0d4f0310-21b2-4304-b0d1-4b482e29af0c.png)
![image](https://user-images.githubusercontent.com/3595443/233968705-d9b766fc-2d21-4259-be91-48e4a81506bc.png)

There should be far more properties and methods in my ViewModel...

**Anyone know why everything has been trimmed? and how to prevent that?
May be compiled bindings? however I am not sure how to use them.**

I tried to add `<AvaloniaUseCompiledBindingsByDefault>true</AvaloniaUseCompiledBindingsByDefault>` in the project containing the View/ViewModel, but then I got that error:

```
Avalonia error AVLN:0004: Cannot parse a compiled binding without
an explicit x:DataType directive to give a starting data type for bindings.
```
**How to fix that? Can we tell the compiler to Trim dotnet framework and third party assemblies, but not our own code?**

3) In another project, I create binding in my code using obj.Bind(property, new Binding(...)) but I got warnings
![image](https://user-images.githubusercontent.com/3595443/233974249-07f7c422-81be-446e-b1a0-a6205ec11ed7.png)

The advice is to use compiled bindings instead, but I cannot found an example how to use them in c# code.
**Anyone have a code snippet for that?**

Thank for any help regarding these topics.

**Comment by timunie** [comment, marked as answer]
See: https://docs.avaloniaui.net/docs/data-binding/compiledbindings 

Compiled Bindings in code is a hard task. I'd rahter add a roots.xml and exclude the needed memebers from trimming.

**Reply by dbriard** [reply]
> See: https://docs.avaloniaui.net/docs/data-binding/compiledbindings
> 
> Compiled Bindings in code is a hard task. I'd rahter add a roots.xml and exclude the needed memebers from trimming.

@timunie Ah, thanks for the tips, Roots.xml make it work!
However, when I created the solution using avalonia template for cross platform, there was already a Roots.xml in the shared project, but not in the Browser project.
According to my tests, Roots.xml seems to to nothing when it is in the shared project, but as soon as I added it to the Browser project, it worked and my ViewModel was not trimmed anymore.

I will do more tests, but I think that should solve most of my issues.
